python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -181,3 +208,77 @@ def __contains__(self, key: _T) -> bool:
+                continue
+            # find the connection corresponding to the input ref
+            connectionRefs = node.quantum.inputs.get(ref.datasetType)
+            # Ref must be an initInput ref and we want to ignore those
+            if connectionRefs is None:
+                continue
+            tmpRefs = set(connectionRefs).difference((ref,))
+            tmpConnections = NamedKeyDict(node.quantum.inputs.items())
+            tmpConnections[ref.datasetType] = list(tmpRefs)
+            helper = AdjustQuantumHelper(inputs=tmpConnections, outputs=node.quantum.outputs)
+            # assert to make the type checker happy, it should not actually be
+            # possible to not have dataId set to None at this point
+            assert node.quantum.dataId is not None
+            # Try to adjust the quantum with the reduced refs to make sure the
+            # node will still satisfy all its conditions.
+            #
+            # If it cant because NoWorkFound is raise, that means a connection
+            # is no longer present, and the node should be removed from the
+            # graph
+            try:
+                helper.adjust_in_place(node.taskDef.connections, node.taskDef.label, node.quantum.dataId)
+                newQuantum = Quantum(taskName=node.quantum.taskName, taskClass=node.quantum.taskClass,
+                                     dataId=node.quantum.dataId, initInputs=node.quantum.initInputs,
+                                     inputs=helper.inputs, outputs=helper.outputs)
+                # If the inputs or outputs were adjusted than the graph builder
+                # thought, dissassociate node from those refs, and if they are
+                # output refs, prune them from downstream tasks
+                for condition, existingMapping, newMapping, remover in (
+                        (helper.inputs_adjusted, node.quantum.inputs, helper.inputs,
+                         datasetRefDict.removeConsumer),
+                        (helper.outputs_adjusted, node.quantum.outputs, helper.outputs,
+                         datasetRefDict.removeProducer)):
+                    if condition:
+                        notNeeded = set()
+                        for key in existingMapping:
+                            if key not in newMapping:
+                                notNeeded |= existingMapping[key]
+                                continue
+                            notNeeded |= set(existingMapping[key]) - set(newMapping[key])
+                        if notNeeded:
+                            for ref in notNeeded:
+                                remover(ref, node)
+                            if remover is datasetRefDict.removeProducer:
+                                _pruner(datasetRefDict, notNeeded, prunes=prunes)
+                object.__setattr__(node, 'quantum', newQuantum)
+
+            except NoWorkFound:
+                # This will throw if the length is less than the minimum number
```

Comment:
Might be simpler to treat this the same as `helper.outputs_adjusted == node.quantum.outputs` (i.e. all outputs have gone away).  I think they're the same.  Might be another thing `AdjustQuantumHelper` should be doing, at least optionally.

Reply:
I am not quite sure what you are trying to say here. im not sure the helper will ever do anything, because an exception was thrown

Reply:
Right; I was just thinking of adding a `try/except` block inside `AdjustQuantumHelper`.  But as per other comment thread, that's not going to happen on this ticket, so I'm just resolving this comment.